Libc/man/utmp.5
.\"	$NetBSD: utmp.5,v 1.5 1994/11/30 19:31:35 jtc Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)utmp.5	8.2 (Berkeley) 3/17/94
.\"
.Dd March 17, 1994
.Dt UTMP 5
.Os BSD 4
.Sh NAME
.Nm utmp ,
.Nm wtmp ,
.Nm lastlog
.Nd login records (DEPRECATED)
.Sh SYNOPSIS
.Fd #include <utmp.h>
.Sh DESCRIPTION
The interfaces in file
.Aq Pa utmp.h
are all
.Sy DEPRECIATED
and are only provided for compatibility with previous releases of Mac OS X.
See
.Xr pututxline 3
and
.Xr utmpx 5
for the supported interfaces.
.Pp
.Aq Pa utmp.h
declares the structures used to record information about current
users in the file
.Nm utmp ,
logins and logouts in the file
.Nm wtmp ,
and last logins in the file
.Nm lastlog .
The time stamps of date changes, shutdowns and reboots are also logged in
the
.Nm wtmp
file.
.Pp
These files can grow rapidly on busy systems, daily or weekly rotation
is recommended. 
If any of these files do not exist, it is not created.
These
files must be created manually and are normally maintained in either the script
.Pa /etc/daily
or the script
.Pa /etc/weekly .
(See
.Xr cron 8 . )
.Bd -literal -offset indent
#define _PATH_UTMP      "/var/run/utmp"
#define _PATH_WTMP      "/var/log/wtmp"
#define _PATH_LASTLOG   "/var/log/lastlog"

#define UT_NAMESIZE     8
#define UT_LINESIZE     8
#define UT_HOSTSIZE     16

struct lastlog {
        time_t  ll_time;
        char    ll_line[UT_LINESIZE];
        char    ll_host[UT_HOSTSIZE];
};

struct utmp {
        char    ut_line[UT_LINESIZE];
        char    ut_name[UT_NAMESIZE];
        char    ut_host[UT_HOSTSIZE];
        time_t	ut_time;
};
.Ed
.Pp
Each time a user logs in, the
.Xr login
program looks up the user's
.Tn UID
in the file
.Nm lastlog.
If it is found, the timestamp of the last time the user logged
in, the terminal line and the hostname
are written to the standard output. (Providing the login is not
.Em quiet ,
see
.Xr login 1 . )
The
.Xr login
program then records the new login time in the file
.Nm lastlog .
.Pp
After the new
.Fa lastlog
record is written ,
.\" the
.\" .Xr libutil 3
.\" routine
the file
.Nm utmp
is opened and the
.Fa utmp
record for the user inserted.
This record remains there until
the user logs out at which time it is deleted.
The
.Nm utmp
file is used by the programs
.Xr rwho 1 ,
.Xr users 1 ,
.Xr w 1 ,
and
.Xr who 1 .
.Pp
Next, the
.Xr login
program opens the file
.Nm wtmp ,
and appends the user's
.Fa utmp
record.
The same
.Fa utmp
record, with an updated time stamp is later appended
to the file when the user logs out. (See
.Xr launchd 8 . )
The
.Nm wtmp
file is used by the programs
.Xr last 1
and
.Xr ac 8 .
.Pp
In the event of a date change, a shutdown or reboot, the
following items are logged in the
.Nm wtmp
file.
.Pp
.Bl -tag -width shutdownxx -compact
.It Li reboot
.It Li shutdown
A system reboot or shutdown has been initiated.
The character
.Ql \&~
is placed in the field
.Fa ut_line ,
and
.Li reboot
or
.Li shutdown
in the field
.Fa ut_name .
(See
.Xr shutdown 8
and
.Xr reboot 8 . )
.Pp
.It Li date
The system time has been manually or automatically updated.
(See
.Xr date 1 . )
The command name
.Xr date
is recorded in the field
.Fa ut_name .
In the field
.Fa ut_line ,
the character
.Ql \\*(Ba
indicates the time prior to the change, and the character
.Ql \&{
indicates the new time.
.El
.Sh FILES
.Bf Sy
(These files no longer exist in 10.5 or later.)
.Ef
.Pp
.Bl -tag -width /var/log/lastlog -compact
.It Pa /var/run/utmp
The
.Nm utmp file.
.It Pa /var/log/wtmp
The
.Nm wtmp file.
.It Pa /var/log/lastlog
The
.Nm lastlog file.
.El
.Sh SEE ALSO
.Xr last 1 ,
.Xr login 1 ,
.Xr who 1 ,
.Xr ac 8 ,
.Xr launchd 8
.Sh HISTORY
A
.Nm utmp
and
.Nm wtmp
file format appeared in
.At v6 .
The
.Nm lastlog
file format appeared in
.Bx 3.0 .


X11/ChangeLog
2006-09-08  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_44

2006-09-06  John Harper  <jsh@apple.com>

	* xc/lib/GL/apple/dri_dispatch.h,
	xc/lib/GL/apple/dri_dispatch.defs: removed glPointParameteri[v]
	aliases

	* xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/applewmstr.h,
	xc/programs/Xserver/hw/apple/applewm.h,
	xc/programs/Xserver/hw/apple/applewm.c,
	xc/programs/Xserver/hw/apple/applewm-impl.h,
	xc/programs/Xserver/hw/apple/applewm-impl.c,
	xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: use shared
	version of AppleWM headers

	* make-host-def, build.sh, Makefile: removed Xplugin/quartz-wm
	references; attempt to support x86_64

2006-07-07  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_43

	* xc/programs/Xserver/hw/apple/x-hash.c: fix compiler warning

	* xc/programs/xterm/main.c: fix xterm build failure (4607761)

2006-02-12  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/GL/apple/indirect.c
	(glAquaInitVisualConfigs): add stereo visuals (4377346)

2006-01-16  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_42

	* xc/programs/Xserver/hw/apple/bundle-main.c: added /sbin and
	/usr/sbin to the default PATH value (4408133)

	* xc/programs/Xserver/hw/apple/quartz.c (do_exec): use same
	default PATH and MANPATH values as bundle-main.c (4408133)

2005-11-29  John Harper  <jsh@apple.com>

	Fix xscreensaver-demo crash and make its cross-process GL
	surfaces work correctly (4356702):

	* xc/programs/Xserver/hw/apple/dri-surface.c: don't compile the
	code to draw on surfaces (that was always disables but was
	causing a crash - I've fixed the crash, but the code is never
	used so best to just remove it from the binary)

2005-10-25  John Harper  <jsh@apple.com>

	* Tagged: LEOPARD_SUBMISSION_0_41

	* xc/lib/xtrans/Xtranssock.c: fix Leopard build failure
	(4222517)

2005-10-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix default
	MANPATH (4312329)

2005-08-29  John Harper  <jsh@apple.com>

	* xc/extras/freetype2/src/base/ftmac.c: merged endian changes
	from 3477131

2005-06-22  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: merged
	fix for CAN-2005-0605 from xfree86 cvs (4031381)

2005-05-24  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c
	(install_ipaddr_source): changed the way we listen for network
	/ hostname changes, hopefully to be more reliable

	* xc/programs/Xserver/hw/apple/X11Application.m (sendEvent):
	change fix for bring-all-to-front not to require private
	headers - hardcode enum value we need :-(  (4075093)

2005-05-23  John Harper  <jsh@apple.com>

	* Makefile, make-host-def, merge-libs.sh,
	xc/config/cf/Imake.cf, xc/config/cf/darwin.cf,
	xc/config/cf/darwinLib.tmpl, xc/include/Xmd.h,
	xc/lib/Xt/TMaction.c, xc/lib/Xt/Xtos.h,
	xc/lib/freetype2/Imakefile: checked in changes to
	support building fat libraries when ppc64 is in RC_ARCHS
	(3721701)

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	bumped short version number to 1.2 so we don't fail to verify

	* xc/programs/Xserver/hw/apple/X11Application.m: handle scroll
	wheel events with delta < 1 correctly (4109092)

2005-05-20  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/rootless.h,
	xc/programs/Xserver/hw/apple/rootless-window.c,
	xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Application.m: recognize the
	bfCPSOrderAllWindowsForward bit in activation events and
	respond to it appropriately (4075093)

2005-03-09  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_40

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-03  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_39

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_38

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_37

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-24  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36_1

	* Makefile: strip-libraries target doesn't strip libXplugin

2004-11-23  John Harper  <jsh@apple.com>

	* Tagged: TIGER_SUBMISSION_0_36

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)


samba/samba/source/smbd/utmp.c
/* 
   Unix SMB/CIFS implementation.
   utmp routines
   Copyright (C) T.D.Lee@durham.ac.uk 1999
   Heavily modified by Andrew Bartlett and Tridge, April 2001
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"

/****************************************************************************
Reflect connection status in utmp/wtmp files.
	T.D.Lee@durham.ac.uk  September 1999

	With grateful thanks since then to many who have helped port it to
	different operating systems.  The variety of OS quirks thereby
	uncovered is amazing...

Hints for porting:
	o  Always attempt to use programmatic interface (pututline() etc.)
	   Indeed, at present only programmatic use is supported.
	o  The only currently supported programmatic interface to "wtmp{,x}"
	   is through "updwtmp*()" routines.
	o  The "x" (utmpx/wtmpx; HAVE_UTMPX_H) seems preferable.
	o  The HAVE_* items should identify supported features.
	o  If at all possible, avoid "if defined(MY-OS)" constructions.

OS observations and status:
	Almost every OS seems to have its own quirks.

	Solaris 2.x:
		Tested on 2.6 and 2.7; should be OK on other flavours.
	AIX:
		Apparently has utmpx.h but doesn't implement.
	OSF:
		Has utmpx.h, but (e.g.) no "getutmpx()".  (Is this like AIX ?)
	Redhat 6:
		utmpx.h seems not to set default filenames.  non-x better.
	IRIX 6.5:
		Not tested.  Appears to have "x".
	HP-UX 9.x:
		Not tested.  Appears to lack "x".
	HP-UX 10.x:
		Not tested.
		"updwtmp*()" routines seem absent, so no current wtmp* support.
		Has "ut_addr": probably trivial to implement (although remember
		that IPv6 is coming...).

	FreeBSD:
		No "putut*()" type of interface.
		No "ut_type" and associated defines. 
		Write files directly.  Alternatively use its login(3)/logout(3).
	SunOS 4:
		Not tested.  Resembles FreeBSD, but no login()/logout().

lastlog:
	Should "lastlog" files, if any, be updated?
	BSD systems (SunOS 4, FreeBSD):
		o  Prominent mention on man pages.
	System-V (e.g. Solaris 2):
		o  No mention on man pages, even under "man -k".
		o  Has a "/var/adm/lastlog" file, but pututxline() etc. seem
		   not to touch it.
		o  Despite downplaying (above), nevertheless has <lastlog.h>.
	So perhaps UN*X "lastlog" facility is intended for tty/terminal only?

Notes:
	Each connection requires a small number (starting at 0, working up)
	to represent the line.  This must be unique within and across all
	smbd processes.  It is the 'id_num' from Samba's session.c code.

	The 4 byte 'ut_id' component is vital to distinguish connections,
	of which there could be several hundred or even thousand.
	Entries seem to be printable characters, with optional NULL pads.

	We need to be distinct from other entries in utmp/wtmp.

	Observed things: therefore avoid them.  Add to this list please.
	From Solaris 2.x (because that's what I have):
		'sN'	: run-levels; N: [0-9]
		'co'	: console
		'CC'	: arbitrary things;  C: [a-z]
		'rXNN'	: rlogin;  N: [0-9]; X: [0-9a-z]
		'tXNN'	: rlogin;  N: [0-9]; X: [0-9a-z]
		'/NNN'	: Solaris CDE
		'ftpZ'	: ftp (Z is the number 255, aka 0377, aka 0xff)
	Mostly a record uses the same 'ut_id' in both "utmp" and "wtmp",
	but differences have been seen.

	Arbitrarily I have chosen to use a distinctive 'SM' for the
	first two bytes.

	The remaining two bytes encode the session 'id_num' (see above).
	Our caller (session.c) should note our 16-bit limitation.

****************************************************************************/

#ifndef WITH_UTMP
/*
 * Not WITH_UTMP?  Simply supply dummy routines.
 */

void sys_utmp_claim(const char *username, const char *hostname, 
		    struct in_addr *ipaddr,
		    const char *id_str, int id_num)
{}

void sys_utmp_yield(const char *username, const char *hostname, 
		    struct in_addr *ipaddr,
		    const char *id_str, int id_num)
{}

#else /* WITH_UTMP */

#ifndef __APPLE__
#include <utmp.h>
#endif

#ifdef HAVE_UTMPX_H
#include <utmpx.h>
#endif

#if defined(__APPLE__)

/* force HAVE_UT_UT_ID so that ut_id_encode() is defined */
#ifndef HAVE_UT_UT_ID
#define HAVE_UT_UT_ID
#endif

static int ut_id_encode(int i, char *fourbyte);

static void
utxcommon(const char *username, const char *hostname,
                   const char *id_str, int id_num, int type)
{
       struct utmpx u;
       bzero(&u, sizeof(u));
       strncpy(u.ut_user, username, sizeof(u.ut_user));
       strncpy(u.ut_host, hostname, sizeof(u.ut_host));
       strncpy(u.ut_line, id_str, sizeof(u.ut_line));
       ut_id_encode(id_num, u.ut_id);
       u.ut_type = type | UTMPX_AUTOFILL_MASK;
       pututxline(&u);
}

void sys_utmp_claim(const char *username, const char *hostname,
                   struct in_addr *ipaddr,
                   const char *id_str, int id_num)
{
       utxcommon(username, hostname, id_str, id_num, USER_PROCESS);
}

void sys_utmp_yield(const char *username, const char *hostname,
                   struct in_addr *ipaddr,
                   const char *id_str, int id_num)
{
       utxcommon(username, hostname, id_str, id_num, DEAD_PROCESS);
}

#else /* !__APPLE__ */

/* BSD systems: some may need lastlog.h (SunOS 4), some may not (FreeBSD) */
/* Some System-V systems (e.g. Solaris 2) declare this too. */
#ifdef HAVE_LASTLOG_H
#include <lastlog.h>
#endif

/****************************************************************************
 Default paths to various {u,w}tmp{,x} files.
****************************************************************************/

#ifdef	HAVE_UTMPX_H

static const char *ux_pathname =
# if defined (UTMPX_FILE)
	UTMPX_FILE ;
# elif defined (_UTMPX_FILE)
	_UTMPX_FILE ;
# elif defined (_PATH_UTMPX)
	_PATH_UTMPX ;
# else
	"" ;
# endif

static const char *wx_pathname =
# if defined (WTMPX_FILE)
	WTMPX_FILE ;
# elif defined (_WTMPX_FILE)
	_WTMPX_FILE ;
# elif defined (_PATH_WTMPX)
	_PATH_WTMPX ;
# else
	"" ;
# endif

#endif	/* HAVE_UTMPX_H */

static const char *ut_pathname =
# if defined (UTMP_FILE)
	UTMP_FILE ;
# elif defined (_UTMP_FILE)
	_UTMP_FILE ;
# elif defined (_PATH_UTMP)
	_PATH_UTMP ;
# else
	"" ;
# endif

static const char *wt_pathname =
# if defined (WTMP_FILE)
	WTMP_FILE ;
# elif defined (_WTMP_FILE)
	_WTMP_FILE ;
# elif defined (_PATH_WTMP)
	_PATH_WTMP ;
# else
	"" ;
# endif

/* BSD-like systems might want "lastlog" support. */
/* *** Not yet implemented */
#ifndef HAVE_PUTUTLINE		/* see "pututline_my()" */
static const char *ll_pathname =
# if defined (_PATH_LASTLOG)	/* what other names (if any?) */
	_PATH_LASTLOG ;
# else
	"" ;
# endif	/* _PATH_LASTLOG */
#endif	/* HAVE_PUTUTLINE */

/*
 * Get name of {u,w}tmp{,x} file.
 *	return: fname contains filename
 *		Possibly empty if this code not yet ported to this system.
 *
 * utmp{,x}:  try "utmp dir", then default (a define)
 * wtmp{,x}:  try "wtmp dir", then "utmp dir", then default (a define)
 */
static void uw_pathname(pstring fname, const char *uw_name, const char *uw_default)
{
	pstring dirname;

	pstrcpy(dirname, "");

	/* For w-files, first look for explicit "wtmp dir" */
	if (uw_name[0] == 'w') {
		pstrcpy(dirname,lp_wtmpdir());
		trim_char(dirname,'\0','/');
	}

	/* For u-files and non-explicit w-dir, look for "utmp dir" */
	if (dirname == 0 || strlen(dirname) == 0) {
		pstrcpy(dirname,lp_utmpdir());
		trim_char(dirname,'\0','/');
	}

	/* If explicit directory above, use it */
	if (dirname != 0 && strlen(dirname) != 0) {
		pstrcpy(fname, dirname);
		pstrcat(fname, "/");
		pstrcat(fname, uw_name);
		return;
	}

	/* No explicit directory: attempt to use default paths */
	if (strlen(uw_default) == 0) {
		/* No explicit setting, no known default.
		 * Has it yet been ported to this OS?
		 */
		DEBUG(2,("uw_pathname: unable to determine pathname\n"));
	}
	pstrcpy(fname, uw_default);
}

#ifndef HAVE_PUTUTLINE

/****************************************************************************
 Update utmp file directly.  No subroutine interface: probably a BSD system.
****************************************************************************/

static void pututline_my(pstring uname, struct utmp *u, BOOL claim)
{
	DEBUG(1,("pututline_my: not yet implemented\n"));
	/* BSD implementor: may want to consider (or not) adjusting "lastlog" */
}
#endif /* HAVE_PUTUTLINE */

#ifndef HAVE_UPDWTMP

/****************************************************************************
 Update wtmp file directly.  No subroutine interface: probably a BSD system.
 Credit: Michail Vidiassov <master@iaas.msu.ru>
****************************************************************************/

static void updwtmp_my(pstring wname, struct utmp *u, BOOL claim)
{
	int fd;
	struct stat buf;

	if (! claim) {
		/*
	 	 * BSD-like systems:
		 *	may use empty ut_name to distinguish a logout record.
		 *
		 * May need "if defined(SUNOS4)" etc. around some of these,
		 * but try to avoid if possible.
		 *
		 * SunOS 4:
		 *	man page indicates ut_name and ut_host both NULL
		 * FreeBSD 4.0:
		 *	man page appears not to specify (hints non-NULL)
		 *	A correspondent suggest at least ut_name should be NULL
		 */
#if defined(HAVE_UT_UT_NAME)
		memset((char *)&u->ut_name, '\0', sizeof(u->ut_name));
#endif
#if defined(HAVE_UT_UT_HOST)
		memset((char *)&u->ut_host, '\0', sizeof(u->ut_host));
#endif
	}
	/* Stolen from logwtmp function in libutil.
	 * May be more locking/blocking is needed?
	 */
	if ((fd = open(wname, O_WRONLY|O_APPEND, 0)) < 0)
		return;
	if (fstat(fd, &buf) == 0) {
		if (write(fd, (char *)u, sizeof(struct utmp)) != sizeof(struct utmp))
		(void) ftruncate(fd, buf.st_size);
	}
	(void) close(fd);
}
#endif /* HAVE_UPDWTMP */

/****************************************************************************
 Update via utmp/wtmp (not utmpx/wtmpx).
****************************************************************************/

static void utmp_nox_update(struct utmp *u, BOOL claim)
{
	pstring uname, wname;
#if defined(PUTUTLINE_RETURNS_UTMP)
	struct utmp *urc;
#endif /* PUTUTLINE_RETURNS_UTMP */

	uw_pathname(uname, "utmp", ut_pathname);
	DEBUG(2,("utmp_nox_update: uname:%s\n", uname));

#ifdef HAVE_PUTUTLINE
	if (strlen(uname) != 0) {
		utmpname(uname);
	}

# if defined(PUTUTLINE_RETURNS_UTMP)
	setutent();
	urc = pututline(u);
	endutent();
	if (urc == NULL) {
		DEBUG(2,("utmp_nox_update: pututline() failed\n"));
		return;
	}
# else	/* PUTUTLINE_RETURNS_UTMP */
	setutent();
	pututline(u);
	endutent();
# endif	/* PUTUTLINE_RETURNS_UTMP */

#else	/* HAVE_PUTUTLINE */
	if (strlen(uname) != 0) {
		pututline_my(uname, u, claim);
	}
#endif /* HAVE_PUTUTLINE */

	uw_pathname(wname, "wtmp", wt_pathname);
	DEBUG(2,("utmp_nox_update: wname:%s\n", wname));
	if (strlen(wname) != 0) {
#ifdef HAVE_UPDWTMP
		updwtmp(wname, u);
		/*
		 * updwtmp() and the newer updwtmpx() may be unsymmetrical.
		 * At least one OS, Solaris 2.x declares the former in the
		 * "utmpx" (latter) file and context.
		 * In the Solaris case this is irrelevant: it has both and
		 * we always prefer the "x" case, so doesn't come here.
		 * But are there other systems, with no "x", which lack
		 * updwtmp() perhaps?
		 */
#else
		updwtmp_my(wname, u, claim);
#endif /* HAVE_UPDWTMP */
	}
}

/****************************************************************************
 Copy a string in the utmp structure.
****************************************************************************/

static void utmp_strcpy(char *dest, const char *src, size_t n)
{
	size_t len = 0;

	memset(dest, '\0', n);
	if (src)
		len = strlen(src);
	if (len >= n) {
		memcpy(dest, src, n);
	} else {
		if (len)
			memcpy(dest, src, len);
	}
}

/****************************************************************************
 Update via utmpx/wtmpx (preferred) or via utmp/wtmp.
****************************************************************************/

static void sys_utmp_update(struct utmp *u, const char *hostname, BOOL claim)
{
#if !defined(HAVE_UTMPX_H)
	/* No utmpx stuff.  Drop to non-x stuff */
	utmp_nox_update(u, claim);
#elif !defined(HAVE_PUTUTXLINE)
	/* Odd.  Have utmpx.h but no "pututxline()".  Drop to non-x stuff */
	DEBUG(1,("utmp_update: have utmpx.h but no pututxline() function\n"));
	utmp_nox_update(u, claim);
#elif !defined(HAVE_GETUTMPX)
	/* Odd.  Have utmpx.h but no "getutmpx()".  Drop to non-x stuff */
	DEBUG(1,("utmp_update: have utmpx.h but no getutmpx() function\n"));
	utmp_nox_update(u, claim);
#else
	pstring uname, wname;
	struct utmpx ux, *uxrc;

	getutmpx(u, &ux);

#if defined(HAVE_UX_UT_SYSLEN)
	if (hostname)
		ux.ut_syslen = strlen(hostname) + 1;	/* include end NULL */
	else
		ux.ut_syslen = 0;
#endif
#if defined(HAVE_UT_UT_HOST)
	utmp_strcpy(ux.ut_host, hostname, sizeof(ux.ut_host));
#endif

	uw_pathname(uname, "utmpx", ux_pathname);
	uw_pathname(wname, "wtmpx", wx_pathname);
	DEBUG(2,("utmp_update: uname:%s wname:%s\n", uname, wname));
	/*
	 * Check for either uname or wname being empty.
	 * Some systems, such as Redhat 6, have a "utmpx.h" which doesn't
	 * define default filenames.
	 * Also, our local installation has not provided an override.
	 * Drop to non-x method.  (E.g. RH6 has good defaults in "utmp.h".)
	 */
	if ((strlen(uname) == 0) || (strlen(wname) == 0)) {
		utmp_nox_update(u, claim);
	} else {
		utmpxname(uname);
		setutxent();
		uxrc = pututxline(&ux);
		endutxent();
		if (uxrc == NULL) {
			DEBUG(2,("utmp_update: pututxline() failed\n"));
			return;
		}
		updwtmpx(wname, &ux);
	}
#endif /* HAVE_UTMPX_H */
}

#endif /* __APPLE__ */

#if defined(HAVE_UT_UT_ID)
/****************************************************************************
 Encode the unique connection number into "ut_id".
****************************************************************************/

static int ut_id_encode(int i, char *fourbyte)
{
	int nbase;
	const char *ut_id_encstr = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
	
	fourbyte[0] = 'S';
	fourbyte[1] = 'M';

/*
 * Encode remaining 2 bytes from 'i'.
 * 'ut_id_encstr' is the character set on which modulo arithmetic is done.
 * Example: digits would produce the base-10 numbers from '001'.
 */
	nbase = strlen(ut_id_encstr);

	fourbyte[3] = ut_id_encstr[i % nbase];
	i /= nbase;
	fourbyte[2] = ut_id_encstr[i % nbase];
	i /= nbase;

	return(i);	/* 0: good; else overflow */
}
#endif /* defined(HAVE_UT_UT_ID) */


#ifndef __APPLE__

/*
  fill a system utmp structure given all the info we can gather 
*/
static BOOL sys_utmp_fill(struct utmp *u,
			  const char *username, const char *hostname,
			  struct in_addr *ipaddr,
			  const char *id_str, int id_num)
{			  
	struct timeval timeval;

	/*
	 * ut_name, ut_user:
	 *	Several (all?) systems seems to define one as the other.
	 *	It is easier and clearer simply to let the following take its course,
	 *	rather than to try to detect and optimise.
	 */
#if defined(HAVE_UT_UT_USER)
	utmp_strcpy(u->ut_user, username, sizeof(u->ut_user));
#elif defined(HAVE_UT_UT_NAME)
	utmp_strcpy(u->ut_name, username, sizeof(u->ut_name));
#endif

	/*
	 * ut_line:
	 *	If size limit proves troublesome, then perhaps use "ut_id_encode()".
	 */
	if (strlen(id_str) > sizeof(u->ut_line)) {
		DEBUG(1,("id_str [%s] is too long for %lu char utmp field\n",
			 id_str, (unsigned long)sizeof(u->ut_line)));
		return False;
	}
	utmp_strcpy(u->ut_line, id_str, sizeof(u->ut_line));

#if defined(HAVE_UT_UT_PID)
	u->ut_pid = sys_getpid();
#endif

/*
 * ut_time, ut_tv: 
 *	Some have one, some the other.  Many have both, but defined (aliased).
 *	It is easier and clearer simply to let the following take its course.
 *	But note that we do the more precise ut_tv as the final assignment.
 */
#if defined(HAVE_UT_UT_TIME)
	GetTimeOfDay(&timeval);
	u->ut_time = timeval.tv_sec;
#elif defined(HAVE_UT_UT_TV)
	GetTimeOfDay(&timeval);
	u->ut_tv = timeval;
#else
#error "with-utmp must have UT_TIME or UT_TV"
#endif

#if defined(HAVE_UT_UT_HOST)
	utmp_strcpy(u->ut_host, hostname, sizeof(u->ut_host));
#endif
#if defined(HAVE_UT_UT_ADDR)
	if (ipaddr)
		u->ut_addr = ipaddr->s_addr;
	/*
	 * "(unsigned long) ut_addr" apparently exists on at least HP-UX 10.20.
	 * Volunteer to implement, please ...
	 */
#endif

#if defined(HAVE_UT_UT_ID)
	if (ut_id_encode(id_num, u->ut_id) != 0) {
		DEBUG(1,("utmp_fill: cannot encode id %d\n", id_num));
		return False;
	}
#endif

	return True;
}

/****************************************************************************
 Close a connection.
****************************************************************************/

void sys_utmp_yield(const char *username, const char *hostname, 
		    struct in_addr *ipaddr,
		    const char *id_str, int id_num)
{
	struct utmp u;

	ZERO_STRUCT(u);

#if defined(HAVE_UT_UT_EXIT)
	u.ut_exit.e_termination = 0;
	u.ut_exit.e_exit = 0;
#endif

#if defined(HAVE_UT_UT_TYPE)
	u.ut_type = DEAD_PROCESS;
#endif

	if (!sys_utmp_fill(&u, username, hostname, ipaddr, id_str, id_num)) return;

	sys_utmp_update(&u, NULL, False);
}

/****************************************************************************
 Claim a entry in whatever utmp system the OS uses.
****************************************************************************/

void sys_utmp_claim(const char *username, const char *hostname, 
		    struct in_addr *ipaddr,
		    const char *id_str, int id_num)
{
	struct utmp u;

	ZERO_STRUCT(u);

#if defined(HAVE_UT_UT_TYPE)
	u.ut_type = USER_PROCESS;
#endif

	if (!sys_utmp_fill(&u, username, hostname, ipaddr, id_str, id_num)) return;

	sys_utmp_update(&u, hostname, True);
}

#endif /* __APPLE__ */
#endif /* WITH_UTMP */
